File: Java/402-perm-java/src/main/java/Perm.java

import java.util.List;

public class Perm {
  public void nextPermutation(List<Integer> perm) {
    int lastIndex = perm.size() - 1;
    int piIndex = lastIndex;
    while (piIndex > 0 && perm.get(piIndex) < perm.get(piIndex - 1)) {
      piIndex--;
    }
    piIndex--;
    int pi = perm.get(piIndex);
    int pjIndex = piIndex + 1;
    for (int i = pjIndex; i <= lastIndex; i++) {
      if (perm.get(i) > pi && perm.get(i) < perm.get(pjIndex)) {
        pjIndex = i;
      }
    }
    int tmp = perm.get(pjIndex);
    perm.set(pjIndex, perm.get(piIndex));
    perm.set(piIndex, tmp);

    for (int i = piIndex + 1, j = lastIndex; i < j; i++, j--) {
      tmp = perm.get(j);
      perm.set(j, perm.get(i));
      perm.set(i, tmp);
    }
  }
}

File: Cpp/408-perm-cpp/src/perm.cc
#include "perm.h"

void NextPermutation(std::vector<int> &perm)
{
    int i = perm.size() - 2;
    while (i >= 0 && perm[i] >= perm[i + 1]) {
        i--;
    }
    if (i >= 0) {
        int j = perm.size() - 1;
        while (j >= 0 && perm[j] <= perm[i]) {
            j--;
        }
        std::swap(perm[i], perm[j]);
    }
    std::reverse(perm.begin() + i + 1, perm.end());
}